src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,179 @@
+        return {};
+
+    const int i = m_msgCount - index.row() + m_startIndex - 1;
+
+    Item *item = m_cache[i];
+    if (!item) {
+        item = new Item{rowData(i)};
+        m_cache.insert(i, item, 1);
+    }
+
+    switch (role) {
+    case Qt::DisplayRole:
+        return item->displayRole;
+    case Qt::ForegroundRole:
+        return item->foregroundRole;
+    case Qt::UserRole:
+        return item->userRole;
+    default:
+        return {};
+    }
+}
+
+void BaseLogModel::addNewMessage(int index)
+{
+    if (m_startIndex == -1)
+        m_startIndex = index;
+
+    beginInsertRows(QModelIndex(), 0, 0);
+    ++m_msgCount;
+    endInsertRows();
+
+    const int count = rowCount();
+    if (count > MAX_LOG_MESSAGES) {
+        const int lastMessage = count - 1;
+        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
+        --m_msgCount;
+        endRemoveRows();
+    }
+}
+
+void BaseLogModel::reset()
+{
+    beginResetModel();
+    m_startIndex = m_msgCount;
+    m_msgCount = 0;
+    endResetModel();
+}
+
+LogMessageModel::LogMessageModel(QObject *parent) : 
+    BaseLogModel(Logger::instance()->messageCount(), parent)
+{
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogMessageModel::handleNewMessage);
+}
+
+void LogMessageModel::handleNewMessage(const Log::Msg &msg)
+{
+    addNewMessage(msg.id);
+}
+
+BaseLogModel::Item LogMessageModel::rowData(const int index) const
+{
+    const Log::Msg msg = Logger::instance()->message(index);
+
+    Item item;
+    const QDateTime time = QDateTime::fromMSecsSinceEpoch(msg.timestamp);
+    item.displayRole = QString::fromLatin1("%1 - %2").arg(time.toString(Qt::SystemLocaleShortDate), msg.message);
+
+    switch (msg.type) {
+    // The RGB QColor constructor is used for performance
+    case Log::NORMAL:
+        item.foregroundRole = QApplication::palette().color(QPalette::WindowText);
+        break;
+
+    case Log::INFO:
+        item.foregroundRole = QColor(0, 0, 255); // blue
+        break;
+
+    case Log::WARNING:
+        item.foregroundRole = QColor(255, 165, 0);  // orange
+        break;
+
+    case Log::CRITICAL:
+        item.foregroundRole = QColor(255, 0, 0);  // red
+        break;
+
+    default:
+        Q_ASSERT(false && "invalid message type");
+    }
+
+    item.userRole = msg.type;
+    return item;
+}
+
+LogPeerModel::LogPeerModel(QObject *parent)
+    : BaseLogModel(Logger::instance()->peerCount(), parent)
+{
+    connect(Logger::instance(), &Logger::newLogPeer, this, &LogPeerModel::handleNewMessage);
+}
+
+void LogPeerModel::handleNewMessage(const Log::Peer &peer)
+{
+    addNewMessage(peer.id);
+}
+
+BaseLogModel::Item LogPeerModel::rowData(const int index) const
+{
+    const Log::Peer peer = Logger::instance()->peer(index);
+
+    const QDateTime time = QDateTime::fromMSecsSinceEpoch(peer.timestamp);
+    QString text = QString::fromLatin1("%1 - ").arg(time.toString(Qt::SystemLocaleShortDate))
```

Comment:
```c++
const QString time = QDateTime::fromMSecsSinceEpoch(peer.timestamp).toString(Qt::SystemLocaleShortDate);
const QString text = QString::fromLatin1("%1 - %2").arg(time, (peer.blocked
        ? tr("%1 was blocked %2", "x.y.z.w was blocked").arg(peer.ip, peer.reason)
        : tr("%1 was banned", "x.y.z.w was banned").arg(peer.ip)));
```

Reply:
done..
extra ident is added by uncrusitify should I keep it?

Reply:
no, it is too long.

Reply:
ok replaced.